README.md
# Octopress Deploy

Easily deploy any static site using S3, Git or Rsync. Pull request to support other deployment methods are welcome.

[![Gem Version](https://badge.fury.io/rb/octopress-deploy.svg)](http://badge.fury.io/rb/octopress-deploy)
[![License](http://img.shields.io/:license-mit-blue.svg)](http://octopress.mit-license.org)

## Installation

Octopress Deploy is bundled with the Octopress Gem, to use it from the command line, install Octopress first.

```
$ gem install octopress
```

## Usage

For usage instructions, please refer to the [Octopress CLI documentation](https://github.com/octopress/octopress#deploy).

## Contributing

1. Fork it
2. Create your feature branch (`git checkout -b my-new-feature`)
3. Commit your changes (`git commit -am 'Add some feature'`)
4. Push to the branch (`git push origin my-new-feature`)
5. Create new Pull Request

## License

Copyright (c) 2014 Brandon Mathis

MIT License

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
"Software"), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


CHANGELOG.md
# Changelog

### unreleased
- New: Files are now uploaded to S3 in parallel. This can be disabled by setting `parallel: false` in `_deploy.yml`. [#63](https://github.com/octopress/deploy/pull/63)

### 1.3.0 - 2015-07-05
- New: Now `_deploy.yml` is processed through ERB, this means you can load configurations for ENV vars pretty easily. [#62](https://github.com/octopress/deploy/pull/62)

### 1.2.8 - 2015-06-29
- Fix: Using Regex quote to escape special characters in repo urls. [#61](https://github.com/octopress/deploy/pull/61)

### 1.2.7 - 2015-06-17
- Fix: Fixed some erroneous references to site_dir configuration. [#60](https://github.com/octopress/deploy/pull/60)
- Minor: Moved documentation to [octopress#deploy](https://github.com/octopress/octopress#deploy).

### 1.2.6 - 2015-06-15
- Fix: Tell users to build their sites if the build dir doesn't exist.

### 1.2.5 - 2015-06-15
- Fix: Ensures that deployment git remote is the same as configuration before pushing.

### 1.2.4 - 2015-05-11
- Fix: `--site` flag now accepts an option as it should.

### 1.2.3 - 2015-05-01
- Minor: Added short flags for --config and --force

### 1.2.2 - 2015-04-28
- Fix: On deploy init git, moved git URL from option to argument since it is required.
- Minor: Now preventing deploy to working branch, complete with helpful error message.

### 1.2.1 - 2015-04-25 

- Fix: Rsync exclude and include options now work properly with multiple items.

### 1.2.0 - 2015-03-07 

- Added Support for Cloudfront invalidation [#49](https://github.com/octopress/deploy/pull/49)

### 1.1.0 - 2015-02-06

- Added incremental uploads to S3. [#46](https://github.com/octopress/deploy/pull/46)

### 1.0.5 - 2015-02-04

- Fixed issue where an S3 pull wouldn't create some directories.
- On S3 deploying, verbose now defaults to false.

### 1.0.4 - 2015-01-17

- Improved help messaging on `deploy init git`.

### 1.0.3 - 2015-01-04

- On add-bucket, added <NAME> as an optional argument instead of an option flag
- Added --force option to pull
- Minor documentation improvements

### 1.0.2 - 2015-01-02

- Updated Octopress docs integration.

### 1.0.1 - 2014-12-27
- Added command line options for setting contents of config file.
- Added support for Octopress docs.

### 1.0.0 - 2014-09-24
- Deemed suitable for use.

###  1.0.0 RC11 - 2014-07-06
- Fix: Changed git add command to be more broadly suitable.

###  1.0.0 RC10 - 2014-07-06
- Fix: Pass options to initialization when adding a bucket. [#34](https://github.com/octopress/deploy/pull/34)

###  1.0.0 RC9 - 2014-05-31

- New: S3 header configuration options. [#25](https://github.com/octopress/deploy/issues/25)
- Fix: `site_dir` defaults to `'_site'`. [#29](https://github.com/octopress/deploy/issues/29)
- Fix: A Windows OS issue. [#30](https://github.com/octopress/deploy/pull/30)

###  1.0.0 RC8 - 2014-05-08

- Now using SafeYAML for loading configurations.

###  1.0.0 RC7 - 2014-05-02

- Fixed: `--version` flag.
- Fixed: Moved requiring `pry-debugger` to tests.

### 1.0.0 RC6 - 2014-04-17

- Fixed: CLI options now override config file settings.

### 1.0.0 RC5 - 2014-04-01

- Fixed: `site_dir` config didn't work.


### 1.0.0 RC4 - 2014-03-25

- Added: Octopress Ink documentation site support

### 1.0.0 RC3 - 2014-03-22

#### Minor Enhancements
- Default rsync flags are now `-rltDvz` so they no
  longer try to set file permissions or preserve owner and
  group name on remote. Issue: #19
- Added `flags` config for rsync allows customization of flags.
- Added support for octopress-ink documentation system.
- Added a `update_docs` Rake task to update assets/docs/index.markdown from README.md.

#### Bug Fixes
- No longer stripping forward slashes on remote_path. Issue #18
- `pull` command no longer promises a default directory.


### 1.0.0 RC2 - 2014-03-18
- CHANGE: `add_bucket` command becomes `add-bucket` 

### 1.0.0 RC1 - 2014-03-17
- Initial release


test/test.rb
require File.expand_path("../../lib/octopress-deploy.rb", __FILE__)
require 'fileutils'
require 'find'
require 'pry-byebug'

@has_failed = false
@failures = {}

def setup_tests
  `rm -rf build` # clean up from previous tests
  generate_site
end

def generate_site
  system "cp -r source/ build"
  Find.find('build').to_a.each do |f| 
    system("echo '#{garbage}' >> #{f}") unless File.directory?(f)
  end
end

def pout(str)
  print str
  $stdout.flush
end

def garbage
  o = [('a'..'z'), ('A'..'Z')].map { |i| i.to_a }.flatten
  (0...50).map { o[rand(o.length)] }.join
end

def diff_dir(dir1, dir2)
  dir_files(dir1).each do |f1|
    if diff = diff_file(f1, f1.sub(dir1, dir2))
      @failures["#{@testing}: #{f1}"] = diff
      pout "F".red
      @has_failed = true
    else
      pout ".".green
    end
  end
end

def diff_file(file1, file2)
  if File.exist?(file2)
    diff = `diff #{file1} #{file2}`
    if diff.size > 0
      diff
    else
      false
    end
  else
    "File: #{file2}: No such file or directory."
  end
end

def dir_files(dir)
  Find.find(dir).to_a.reject!{|f| File.directory?(f) }
end

def test_remote_git
  @testing = 'remote git'
  repo = "git@github.com:octopress/deploy"
  config = "_git_deploy.yml"
  
  # Clean up from previous tests
  #
  `rm -rf .deploy pull-git`
  
  # Test remote git deployment
  #
  Octopress::Deploy.init_config(method: 'git', config_file: config, force: true, site_dir: 'build', git_branch: 'test_git_deploy', git_url: repo)
  Octopress::Deploy.push(config_file: config)
  Octopress::Deploy.pull(dir: 'pull-git', config_file: config)
  diff_dir('build', 'pull-git')
end

def test_local_git
  @testing = 'local git'
  config = "_git_deploy.yml"

  # Clean up from previous tests
  #
  `rm -rf .deploy local-git pull-git`
  `git init --bare local-git`
  repo = "local-git"

  # Test local git deployment
  #
  Octopress::Deploy.init_config(method: 'git', config_file: config, force: true, site_dir: 'build', git_branch: 'test_git_deploy', git_url: File.expand_path(repo), remote_path: 'site')
  Octopress::Deploy.push(config_file: config)
  Octopress::Deploy.pull(dir: 'pull-git', config_file: config, git_url: File.expand_path(repo), remote_path: 'site')
  diff_dir('build', 'pull-git/site')
end

def test_remote_rsync
  @testing = 'remote rsync'
  config = '_rsync_deploy.yml'

  # Clean up from previous tests
  `rm -rf local-rsync pull-rsync`
  
  # Test remote git deployment
  #
  Octopress::Deploy.init_config(method: 'rsync', config_file: config, site_dir: 'build', force: true, user: 'imathis@imathis.com', remote_path: '~/octopress-deploy/rsync/')
  Octopress::Deploy.push(config_file: config)
  Octopress::Deploy.pull(dir: 'pull-rsync', config_file: config)
  diff_dir('build', 'pull-rsync')

end

def test_local_rsync
  @testing = 'local rsync'
  config = '_rsync_deploy.yml'
  
  # Clean up from previous tests
  `rm -rf local-rsync pull-rsync`

  # Test local git deployment
  #
  Octopress::Deploy.init_config(method: 'rsync', config_file: config, site_dir: 'build', force: true, remote_path: 'local-rsync')
  Octopress::Deploy.push(config_file: config)
  Octopress::Deploy.pull(dir: 'pull-rsync', config_file: config, user: false, remote_path: 'local-rsync')
  diff_dir('build', 'pull-rsync')
end

def test_s3
  @testing = 's3'
  config = "_s3_deploy.yml"
  `rm -rf pull-s3`
  Octopress::Deploy.push(config_file: config)
  Octopress::Deploy.pull(dir: 'pull-s3', config_file: config)
  diff_dir('build', 'pull-s3')
end

def print_test_results
  puts "\n"
  if @has_failed
    @failures.each do |name, diff|
      puts "Failure in #{name}:".red
      puts "---------"
      puts diff
      puts "---------"
    end
    abort
  else
    puts "All passed!".green
  end
end


setup_tests

# local tests
test_local_rsync
test_local_git

# remote tests
test_remote_git
test_remote_rsync
test_s3

print_test_results



lib/octopress-deploy.rb
$LOAD_PATH.unshift File.expand_path("../", __FILE__)

require 'colorator'
require 'yaml'
require 'octopress'
require 'safe_yaml/load'

require 'octopress-deploy/version'
require 'octopress-deploy/core_ext'
require 'octopress-deploy/commands'

module Octopress
  module Deploy
    autoload :Git,    'octopress-deploy/git'
    autoload :Rsync,  'octopress-deploy/rsync'
    autoload :S3,     'octopress-deploy/s3'

    METHODS = {
      'git'   => Git,
      'rsync' => Rsync,
      's3'    => S3
    }

    DEFAULT_OPTIONS = {
      :config_file => '_deploy.yml',
      :site_dir => '_site',
    }

    def self.push(options={})
      options = merge_configs(options)
      deployer(options).push
    end

    def self.pull(options={})
      options = merge_configs(options)

      if Dir.exist?(options[:dir]) &&
          !(Dir.entries(options[:dir]) - %w{. ..}).empty? &&
          !options[:force]
            puts "Pull failed. Directory #{options[:dir]} is not empty. Pass --force to overwrite."
            abort
      else
        FileUtils.mkdir_p options[:dir]
        deployer(options).pull
      end
    end

    def self.add_bucket(options={})
      options = merge_configs(options)
      get_deployment_method(options).new(options).add_bucket()
    end

    def self.merge_configs(options={})
      options = check_config(options)
      config  = SafeYAML.load(ERB.new(File.read(options[:config_file])).result).to_symbol_keys
      options = config.deep_merge(options)
    end

    def self.check_config(options={})
      options = options.to_symbol_keys
      options[:config_file] ||= DEFAULT_OPTIONS[:config_file]

      if !File.exists? options[:config_file]
        abort "File not found: #{options[:config_file]}. Create a deployment config file with `octopress deploy init <METHOD>`."
      end

      options
    end

    def self.deployer(options)
      get_deployment_method(options).new(options)
    end

    def self.get_deployment_method(options)
      METHODS[options[:method].downcase]
    end


    def self.site_dir
      if options[:site_dir]
        options[:site_dir]
      elsif File.exist? '_config.yml'
        SafeYAML.load(File.open('_config.yml'))['site_dir'] || '_site'
      else
        '_site'
      end
    end

    # Create a config file
    #
    def self.init_config(options={})
      options = options.to_symbol_keys

      if !options[:method]
        abort "Please provide a deployment method. e.g. #{METHODS.keys}"
      end

      @options = DEFAULT_OPTIONS.deep_merge(options)
      write_config
      check_gitignore
    end

    def self.write_config
      if File.exist?(@options[:config_file]) && !@options[:force]
        abort "A config file already exists at #{@options[:config_file]}. Use --force to overwrite."
      end

      config = get_config.strip
      File.open(@options[:config_file], 'w') { |f| f.write(config) }
      puts "File #{@options[:config_file]} created.".green
      puts "------------------"
      puts "#{config.yellow}"
      puts "------------------"
      puts "Modify these configurations as necessary."
    end

    def self.get_config
      <<-FILE
#{"method: #{@options[:method]}".ljust(40)}  # How do you want to deploy? git, rsync or s3.
#{"site_dir: #{@options[:site_dir]}".ljust(40)}  # Location of your static site files.

#{get_deployment_method(@options).default_config(@options)}
FILE
    end

    # Checks the repository's .gitignore for the config file
    #
    # returns: Boolean - whether it is present or not.
    #
    def self.check_gitignore
      gitignore = File.join(`git rev-parse --show-toplevel`.strip, ".gitignore")

      if !File.exist?(gitignore) ||
        File.open(gitignore).read.match(/^#{@options[:config_file]}/i).nil?
        puts "Remember to add #{@options[:config_file]} to your .gitignore."
        false
      else
        true
      end
    end

    def self.gem_dir(*subdirs)
      File.expand_path(File.join(File.dirname(__FILE__), '../', *subdirs))
    end

  end
end


lib/octopress-deploy/s3.rb
require 'find'
require 'fileutils'

module Octopress
  module Deploy
    class S3

      def initialize(options)
        begin
          require 'aws-sdk-v1'
        rescue LoadError
          abort "Deploying to S3 requires the aws-sdk-v1 gem. Install with `gem install aws-sdk-v1`."
        end
        @options     = options
        @local       = options[:site_dir]          || '_site'
        @bucket_name = options[:bucket_name]
        @access_key  = options[:access_key_id]     || ENV['AWS_ACCESS_KEY_ID']
        @secret_key  = options[:secret_access_key] || ENV['AWS_SECRET_ACCESS_KEY']
        @region      = options[:region]            || ENV['AWS_DEFAULT_REGION'] || 'us-east-1'
        @distro_id   = options[:distribution_id]   || ENV['AWS_DISTRIBUTION_ID']
        @remote_path = (options[:remote_path]      || '/').sub(/^\//,'')
        @verbose     = options[:verbose]
        @incremental = options[:incremental]       || true
        @delete      = options[:delete]
        @headers     = options[:headers]           || []
        @remote_path = @remote_path.sub(/^\//,'')  # remove leading slash
        @pull_dir    = options[:dir]
        @bust_cache_files = []
        @thread_pool = []
        connect
      end

      def push
        #abort "Seriously, you should. Quitting..." unless Deploy.check_gitignore
        @bucket = @s3.buckets[@bucket_name]
        if !@bucket.exists?
          abort "Bucket not found: '#{@bucket_name}'. Check your configuration or create a bucket using: `octopress deploy add-bucket`"
        else
          if File.exist?(@local)
            puts "Syncing #{@local} files to #{@bucket_name} on S3."
            write_files
            delete_files if delete_files?
            status_message
          else
            abort "Cannot find site build at #{@local}. Be sure to build your site first."
          end
        end
      end

      def pull
        @bucket = @s3.buckets[@bucket_name]
        if !@bucket.exists?
          abort "Bucket not found: '#{@bucket_name}'. Check your configuration or create a bucket using: `octopress deploy add-bucket`"
        else
          puts "Syncing from S3 bucket: '#{@bucket_name}' to #{@pull_dir}."
          @bucket.objects.each do |object|
            path = File.join(@pull_dir, object.key)

            # Path is a directory, not a file
            if path =~ /\/$/ 
              FileUtils.mkdir_p(path) unless File.directory?(path)
            else
              dir = File.dirname(path)
              FileUtils.mkdir_p(dir) unless File.directory?(dir)
              File.open(path, 'w') { |f| f.write(object.read) }
            end
          end
        end
      end

      # Connect to S3 using the AWS SDK
      # Retuns an aws bucket
      #
      def connect
        AWS.config(access_key_id: @access_key, secret_access_key: @secret_key, region: @region)
        @s3 = AWS.s3
        @cloudfront = AWS.cloud_front.client
      end

      # Write site files to the selected bucket
      #
      def write_files
        puts "Writing #{pluralize('file', site_files.size)}#{" (sequential mode)" unless parallel_upload?}:" if @verbose
        @bust_cache_files = []

        site_files.each do |file|
          if parallel_upload?
            threaded { write_file file }
          else
            write_file file
          end
        end

        @thread_pool.each(&:join)
        bust_cloudfront_cache
      end

      def write_file file
        if write_file? file
          s3_upload_file file
          @bust_cache_files << file
          @verbose ? puts("+ #{remote_path(file)}") : progress('+')
        else
          @verbose ? puts("= #{remote_path(file)}") : progress('=')
        end
      end

      def s3_upload_file file
        s3_object(file).write File.open(file), s3_object_options(file)
      end

      def bust_cloudfront_cache
        return if @distro_id.nil?

        puts "Invalidating cache for #{pluralize('file', site_files.size)}" if @verbose
        @cloudfront.create_invalidation(
          distribution_id: @distro_id, 
          invalidation_batch:{
            paths:{
              quantity: @bust_cache_files.size,
              items: @bust_cache_files.map{|file| "/" + remote_path(file)}
            },
            # String of 8 random chars to uniquely id this invalidation
            caller_reference: (0...8).map { ('a'..'z').to_a[rand(26)] }.join
          }
        ) unless @bust_cache_files.empty?
        @bust_cache_files = []
      end

      def s3_object_options(file)
        s3_filename = remote_path file
        s3_options = { :acl => :public_read }

        @headers.each do |conf|
          if conf.has_key? 'filename' and s3_filename.match(conf['filename'])
            if @verbose
              puts "+ #{s3_filename} matched pattern #{conf['filename']}"
            end

            if conf.has_key? 'expires'
              expireDate = conf['expires']

              relative_years = /^\+(\d+) year(s)?$/.match(conf['expires'])
              if relative_years
                expireDate = (Time.now + (60 * 60 * 24 * 365 * relative_years[1].to_i)).httpdate
              end

              relative_days = /^\+(\d+) day(s)?$/.match(conf['expires'])
              if relative_days
                expireDate = (Time.now + (60 * 60 * 24 * relative_days[1].to_i)).httpdate
              end

              s3_options[:expires] = expireDate
            end

            if conf.has_key? 'content_type'
              s3_options[:content_type] = conf['content_type']
            end

            if conf.has_key? 'cache_control'
              s3_options[:cache_control] = conf['cache_control']
            end

            if conf.has_key? 'content_encoding'
              s3_options[:content_encoding] = conf['content_encoding']
            end
          end
        end

        s3_options
      end

      # Delete files from the bucket, to ensure a 1:1 match with site files
      #
      def delete_files
        if deletable_files.size > 0
          puts "Deleting #{pluralize('file', deletable_files.size)}:" if @verbose
          deletable_files.each do |file|
            @bucket.objects.delete(file)
            if @verbose
              puts "- #{file}"
            else
              progress('-')
            end
          end
        end
      end

      # Create a new S3 bucket
      #
      def add_bucket
        puts @bucket_name
        @bucket = @s3.buckets.create(@bucket_name)
        puts "Created new bucket '#{@bucket_name}' in region '#{@region}'."
        configure_bucket
      end

      def configure_bucket
        error_page = @options['error_page'] || remote_path('404.html')
        index_page = @options['index_page'] || remote_path('index.html')

        config = @bucket.configure_website do |cfg|
          cfg.index_document_suffix = index_page
          cfg.error_document_key = error_page
        end
        puts "Bucket configured with index_document: #{index_page} and error_document: #{error_page}."
      end

      def delete_files?
        !!@delete
      end

      # local site files
      def site_files
        @site_files ||= Find.find(@local).to_a.reject do |f|
          File.directory?(f)
        end
      end

      # Destination paths for local site files.
      def site_files_dest
        @site_files_dest ||= site_files.map{|f| remote_path(f) }
      end

      # Replace local path with remote path
      def remote_path(file)
        File.join(@remote_path, file.sub(@local, '')).sub(/^\//, '')
      end

      # Files from the bucket which are deletable
      # Only deletes files beneath the remote_path if specified
      def deletable_files
        return [] unless delete_files?
        unless @deletable
          @deletable = @bucket.objects.map(&:key) - site_files_dest
          @deletable.reject!{|f| (f =~ /^#{@remote_path}/).nil? }
        end
        @deletable
      end

      # List written and deleted file counts
      def status_message
        uploaded = site_files.size
        deleted = deletable_files.size

        message =  "\nSuccess:".green + " #{uploaded} #{pluralize('file', uploaded)} uploaded"
        message << ", #{deleted} #{pluralize('file', deleted)} deleted."
        puts message
        configure_bucket unless @bucket.website?
      end

      # Print consecutive characters
      def progress(str)
        print str
        $stdout.flush
      end

      def pluralize(str, num)
        str << 's' if num != 1
        str
      end

      # Return default configuration options for this deployment type
      def self.default_config(options={})
        <<-CONFIG
#{"bucket_name: #{options[:bucket_name]}".ljust(40)}  # Name of the S3 bucket where these files will be stored.
#{"access_key_id: #{options[:access_key_id]}".ljust(40)}  # Get this from your AWS console at aws.amazon.com.
#{"secret_access_key: #{options[:secret_access_key]}".ljust(40)}  # Keep it safe; keep it secret. Keep this file in your .gitignore.
#{"distribution_id: #{options[:distribution_id]}".ljust(40)}  # Get this from your CloudFront page at https://console.aws.amazon.com/cloudfront/
#{"remote_path: #{options[:remote_path] || '/'}".ljust(40)}  # relative path on bucket where files should be copied.
#{"region: #{options[:remote_path] || 'us-east-1'}".ljust(40)}  # Region where your bucket is located.
#{"verbose: #{options[:verbose] || 'false'}".ljust(40)}  # Print out all file operations.
#{"incremental: #{options[:incremental] || 'true'}".ljust(40)}  # Only upload new/changed files
#{"delete: #{options[:delete] || 'false'}".ljust(40)}  # Remove files from destination which do not match source files.
#{"parallel: #{options[:parallel] || 'true'}".ljust(40)}  # Speed up deployment by uploading files in parallel.
CONFIG
      end

    protected

      def write_file? file
        return true unless @incremental

        file_digest = Digest::MD5.file(file).hexdigest
        o = s3_object file
        s3sum = o.etag.tr('"','') if o.exists?
        s3sum.to_s != file_digest
      end

      def s3_object file
        s3_filename = remote_path file
        @bucket.objects[s3_filename]
      end

      def parallel_upload?
        @options[:parallel]
      end

      def threaded &blk
        @thread_pool << Thread.new(blk) do |operation|
          operation.call
        end
      end

    end
  end
end


lib/octopress-deploy/cli.rb
module Octopress
  module Deploy
    class Cli < Octopress::Command
      def self.init_with_program(p)
        p.command(:deploy) do |c|
          c.syntax "octopress deploy [options]"
          c.description "Deploy your Octopress site."
          c.option "using", "--using METHOD", "Define the push method to use, overriding your configuration file's setting"
          c.option "config_file", "--config FILE", "The path to your config file (default: _deploy.yml)"
          c.option "init", "--init METHOD", "Initialize a config file with the options for the given method."
          c.option "pull", "--pull DIRECTORY", "Pull down the published copy of your site into a directory (default: ./site-pull)"

          c.action do |_, options|
            if options["init"] and options["init"].is_a?(String)
              Octopress::Deploy.init_config(options["init"], options)
            elsif options["pull"] and options["pull"].is_a?(String)
              Octopress::Deploy.pull(options["pull"], options)
            else
              Octopress::Deploy.push(options)
            end
          end
        end
      end
    end
  end
end


lib/octopress-deploy/rsync.rb
module Octopress
  module Deploy
    class Rsync

      def initialize(options)
        @options      = options
        @flags        = @options[:flags] || ' -avz'
        @user         = @options[:user]
        @port         = @options[:port]
        @local        = @options[:site_dir] || '_site'
        @remote_path  = @options[:remote_path]
        @exclude      = @options[:exclude]
        @exclude_from = @options[:exclude_from]
        @exclude_from = File.expand_path(@exclude_from) if @exclude_from
        @include      = @options[:include]
        @include_from = @options[:include_from]
        @include_from = File.expand_path(@include_from) if @include_from
        @delete       = @options[:delete] || false
        @pull_dir     = @options[:dir]
      end

      def push
        if File.exist?(@site_dir)
          puts "Syncing #{@local} files to #{@remote_path} with rsync."
          system cmd
        else
          abort "Cannot find site build at #{@local}. Be sure to build your site first."
        end
      end

      def pull
        puts "Syncing #{@remote_path} files to #{@pull_dir} with rsync."
        system cmd
      end

      def cmd
        local = ''
        remote = ''

        cmd    =  "rsync "
        cmd    << "#{@flags} "
        cmd    << " --exclude-from #{@exclude_from}"  if @exclude_from
        Array(@exclude).each do |e|
          cmd  << " --exclude #{e}"
        end
        cmd    << " --include-from #{@include_from}"  if @include_from
        Array(@include).each do |i|
          cmd  << " --include #{i}"
        end
        cmd    << " --rsh='ssh -p#{@port}'"           if @user && @port
        cmd    << " --delete "                        if @delete

        local  << " #{File.join(@local, '')} "
        remote << " #{@user}:"                         if @user
        remote << "#{@remote_path}"

        if @pull_dir
          cmd << remote+'/ ' << @pull_dir
        else
          cmd << local << remote
        end
      end

      def self.default_config(options={})
        <<-CONFIG
#{"user: #{options[:user]}".ljust(40)}  # The user for your host, e.g. user@host.com
#{"remote_path: #{options[:remote_path]}".ljust(40)}  # Destination directory
#{"delete: #{options[:delete]}".ljust(40)}  # Remove files from destination which don't match files in source
#{"port: #{options[:port]}".ljust(40)}  # If your host requires a non standard port
#{"flags: #{options[:flags] || ' -avz'}".ljust(40)}  # Modify flags as necessary to suit your hosting setup

#{"# exclude: ".ljust(40)}  # files to exclude
#{"# exclude-from: ".ljust(40)}  # Path to file containing list of files to exclude
#{"# include: ".ljust(40)}  # files to include
#{"# include-from: ".ljust(40)}  # Path to file containing list of files to include
CONFIG
      end

    end
  end
end


lib/octopress-deploy/commands.rb
module Octopress
  module Deploy
    class Commands < Octopress::Command
      def self.init_with_program(p)
        p.command(:deploy) do |c|
          c.syntax "deploy [options]"
          c.version Deploy::VERSION
          c.description "Deploy your Octopress site."
          c.option "config_file", "-c", "--config FILE", "The path to your config file (default: _deploy.yml)"
          c.option "stage", "-s", "--stage STAGE", "The stage for deployment (default: production)"

          c.action do |args, options|
            Octopress::Deploy.push(options)
          end

          c.command(:pull) do |c|
            c.syntax "pull <DIR>"
            c.description "Pull down the published copy of your site into DIR"
            c.option 'force', '-f', '--force', 'Overwrite existing files on pull'
            c.option "config_file", "-c", "--config FILE", "The path to your config file (default: _deploy.yml)"

            c.action do |args, options|
              options['dir'] = args.first
              Octopress::Deploy.pull(options)
            end
          end

          c.command(:init) do |c|
            c.syntax 'init <METHOD> [options]'
            c.description "Create a configuration file for a deployment method (#{Deploy::METHODS.keys.join(', ')})."

            c.command(:rsync) do |c|
              c.syntax 'rsync [options]'
              c.description "Create an rsync deployment configuration file."
              c.option 'user', '-u', '--user USER', 'SSH user (e.g. username@host.org)'
              c.option 'port', '-p', '--port PORT', 'SSH port (default: 22)'
              c.option 'flags', '-F', '--flags FLAGS', 'Flags to pass to Rsync command (default: -avz)'
              c.option 'remote_path', '--dir DIR', 'Path to site directory on host (e.g. ~/webroot/)'
              c.option 'delete', '--delete', 'Sync file deletion'
              add_common_init_options(c)

              c.action do |args, options|
                options['method'] = 'rsync'
                Octopress::Deploy.init_config(options)
              end
            end

            c.command(:git) do |c|
              c.syntax 'git <URL> [options]'
              c.description "Create a git deployment configuration file."
              c.option 'git_branch', '-b', '--branch NAME', 'Branch to deploy to (default: master)'
              c.option 'remote_path', '-d', '--dir DIR', 'Deploy site into a subdirectory.'
              c.option 'delete', '--delete', 'Sync file deletion'
              add_common_init_options(c)

              c.action do |args, options|
                options['method'] = 'git'
                options['git_url'] = args.first
                if options['git_url']
                  options['git_url'] = options['git_url'].sub(/^\./, Dir.pwd)
                else
                  $stderr.puts "git url missing"
                  puts c # print the help
                  abort
                end
                Octopress::Deploy.init_config(options)
              end
            end

            c.command(:s3) do |c|
              c.syntax 's3 [options]'
              c.description "Create an S3 deployment configuration file."
              c.option 'bucket_name', '-b', '--bucket NAME', 'S3 Bucket name'
              c.option 'access_key_id', '-a', '--access KEY', 'Access key id'
              c.option 'secret_access_key', '-s', '--secret KEY', 'Secret access key'
              c.option 'region', '-r', '--region REGION', 'AWS region (default: us-east-1)'
              c.option 'remote_path', '-d', '--dir DIR', 'Deploy site into a subdirectory.'
              c.option 'verbose', '-v', '--verbose', 'Log verbose output when deploying'
              c.option 'incremental', '-i', '--incremental', 'Only upload new/changed files'
              c.option 'delete', '--delete', 'Sync file deletion'
              add_common_init_options(c)

              c.action do |args, options|
                options['method'] = 's3'
                Octopress::Deploy.init_config(options)
              end
            end
          end

          c.command(:'add-bucket') do |c|
            c.syntax 'add-bucket <NAME> [options]'
            c.description "Add a new S3 bucket and configure it for static websites. Name defaults to bucket_name in config file"
            c.option 'region','--region REGION','Choose a region. (Defaults: to region in config file)'
            c.option 'index_page','--index PAGE','Specify an index page. (Default: index.html)'
            c.option 'error_page','--error PAGE','Specify an error page. (Default: 404.html)'
            c.option "config_file", "-c", "--config FILE", "The path to your config file (default: _deploy.yml)"

            c.action do |args, options|
              options['bucket_name'] = args.first
              Octopress::Deploy.add_bucket(options)
            end
          end
        end
      end

      def self.add_common_init_options(c)
        c.option 'site_dir', '--site DIR', 'Path to generated site (default: _site).'
        c.option 'force', '-f', '--force', 'Overwrite any exiting config file.'
        c.option 'config_file', "-c", '--config FILE', 'Choose a config file name. (default. _deploy.yml)'
      end
    end
  end
end
